pkg/reconciler/base_test.go

```diff
@@ -107,6 +118,191 @@ func TestHasAdapterLabelsForType(t *testing.T) {
+						Controller: ptr.Bool(true),
+					}},
+				}},
+			},
+			expect: &metav1.OwnerReference{
+				APIVersion: "apps/v1",
+				Kind:       "ReplicaSet",
+				Name:       "foo-fake-adapter-abc012",
+			},
+		},
+		/* In this test, the ReplicaSet is owned by a Deployment, but
+		   this Deployment can't be found in the lister so the ancestor
+		   is considered unknown.
+		*/
+		"One supported kind of ancestor doesn't exist": {
+			objects: []runtime.Object{
+				&appsv1.ReplicaSet{ObjectMeta: metav1.ObjectMeta{
+					Namespace: tNs,
+					Name:      "foo-fake-adapter-abc012",
+					OwnerReferences: []metav1.OwnerReference{{
+						APIVersion: appsv1.SchemeGroupVersion.String(),
+						Kind:       "Deployment",
+						Name:       "foo-fake-adapter",
+						Controller: ptr.Bool(true),
+					}},
+				}},
+				&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
+					Namespace: tNs,
+					Name:      "foo-fake-adapter-abc012-efg345",
+					OwnerReferences: []metav1.OwnerReference{{
+						APIVersion: appsv1.SchemeGroupVersion.String(),
+						Kind:       "ReplicaSet",
+						Name:       "foo-fake-adapter-abc012",
+						Controller: ptr.Bool(true),
+					}},
+				}},
+			},
+			expect: nil,
+		},
+		/* In this test, the outermost ancestor of kind Foo can't be
+		   reached because one intermediate ancestor is of kind Service,
+		   which the recursive resolver doesn't support.
+		   The Service is returned as the last resolvable ancestor.
+		*/
+		"One kind of ancestor is not supported for recursive lookup": {
+			objects: []runtime.Object{
+				&corev1.Service{ObjectMeta: metav1.ObjectMeta{
+					Namespace: tNs,
+					Name:      "fake",
+					OwnerReferences: []metav1.OwnerReference{{
+						APIVersion: "fake/v0",
+						Kind:       "Foo",
+						Name:       "fake",
+						Controller: ptr.Bool(true),
+					}},
+				}},
+				&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
+					Namespace: tNs,
+					Name:      "fake",
+					OwnerReferences: []metav1.OwnerReference{{
+						APIVersion: corev1.SchemeGroupVersion.String(),
+						Kind:       "Service",
+						Name:       "fake",
+						Controller: ptr.Bool(true),
+					}},
+				}},
+			},
+			expect: &metav1.OwnerReference{
+				APIVersion: "v1",
+				Kind:       "Service",
+				Name:       "fake",
+			},
+		},
+		/* In this test, the object to resolve the ancestor for doesn't
+		   have a controller. Because it is the initial object, there
+		   is no ancestor to return at all.
+		*/
+		"No controller": {
+			objects: []runtime.Object{
+				&corev1.Pod{ObjectMeta: metav1.ObjectMeta{
+					Namespace:       tNs,
+					Name:            "fake",
+					OwnerReferences: []metav1.OwnerReference{},
+				}},
+			},
+			expect: nil,
+		},
+	}
+
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			withK8sCli := func(ctx context.Context, _ *rest.Config) context.Context {
+				ctx, _ = fakek8sclient.With(ctx, tc.objects...)
+				return ctx
+			}
+			injection.Fake.RegisterClient(withK8sCli)
```

Comment:
@tzununbekov I remember we were struggling to implement a similar test in Knative Eventing (https://github.com/knative/eventing/pull/5880) and resorted to do something super complicated where we injected our own started informer to the context.

I turns out that registering a ClientSet with pre-populated objects is in fact very simple 🙃

Reply:
Nice! It would save me a lot of headache in the upstream issue.